keras/src/quantizers/gptq.py

```diff
@@ -0,0 +1,252 @@
+from keras.src import ops
+from keras.src.layers import Dense
+from keras.src.layers import EinsumDense
+from keras.src.quantizers.gptqquant import dequantize
+
+
+class GPTQ:
+    def __init__(self, layer):
+        self.original_layer = layer
+        self.kernel_shape = layer.kernel.shape
+        self.nsamples = 0
+        self.quantizer = None
+
+        # Explicitly handle each supported layer type
+        if isinstance(layer, Dense) or (
+            isinstance(layer, EinsumDense) and layer.kernel.ndim == 2
+        ):
+            # For a standard Dense layer, the dimensions are straightforward.
+            self.rows = self.kernel_shape[0]  # Input features
+            self.columns = self.kernel_shape[1]  # Output features
+            self.layer = layer  # The layer itself can be used directly.
+
+        # Handle 3D EinsumDense layers (typically from attention blocks).
+        elif isinstance(layer, EinsumDense) and layer.kernel.ndim == 3:
+            # For EinsumDense, we determine the effective 2D dimensions.
+            shape = list(self.kernel_shape)
+            try:
+                d_model_dim_index = shape.index(max(shape))
+            except ValueError:
+                raise TypeError(
+                    f"Could not determine hidden dimension from shape {shape}"
+                )
+
+            if d_model_dim_index == 0:  # QKV projection case
+                in_features, heads, head_dim = shape
+                self.rows, self.columns = in_features, heads * head_dim
+            elif d_model_dim_index in [1, 2]:  # Attention Output case
+                heads, head_dim, out_features = shape
+                self.rows, self.columns = heads * head_dim, out_features
+
+            # Create a temporary object that holds a reshaped
+            # 2D version of the kernel.
+            self.layer = type(
+                "temp",
+                (object,),
+                {
+                    "kernel": ops.reshape(
+                        layer.kernel, (self.rows, self.columns)
+                    ),
+                    "bias": layer.bias,
+                },
+            )()
```

Comment:
for readability sake, should we move this out into a helper (`_handle_3d_einsumdense`)

Reply:
added comments, moving out would again lead to complexity